### DISCUSSION TITLE
Optional field != optional value

### DISCUSSION THREAD

**Discussion by wombatonfire** [pydantic/pydantic#5076] [Pydantic V2]
Hello,

I've been recently reading a lot about the issues and solutions/workarounds related to required/optional/nullable fields. I've also seen the [announcement](https://docs.pydantic.dev/blog/pydantic-v2/#required-vs-nullable-cleanup) about improvements coming in V2. And still, I believe the problem can (should?) be handled in a different way. I thought I'd share my thoughts on this and, if anything of the following makes sense, I would be glad to participate in making things better in this regard.

There are various use cases when dealing with required/optional/nullable fields, leading to slightly different issues and potential solutions, but, when I tried to generalize them and figure out what was the fundamental problem that led to those issues, I came to a conclusion that there are 2 main candidates:

1. The use of annotations/default values to indicate an optional field.
2. The inclusion of the optional fields into the initialized model.

The first problem looks semantics-related to me. Type annotations and default values are data attributes, while fields themselves (keys) are more like metadata. The use of a data attribute (default value) to convey the meaning in the metadata scope doesn't look right. Particularly, because the same attribute is used simultaneously for a different purpose in the data scope. Crossing this data/metadata boundary leads to confusion and asks for refactoring. This is why, for example, users get confused when they want an optional field, but don't want to allow `None` as a data value. Separate concerns are currently covered by the same tool.

I feel that the second problem is somewhat related to the first one. It's also about the metadata, describing the data. The existence/absence of the key (field) is the way of describing the data. Passing the key with a default value, even if the value is `null`/`None`/`0`/empty string, etc. is not semantically the same as not passing the key at all. Unfortunately, Pydantic initializes the model with all the fields, even when they are optional and you don't pass them. One specific use case where that is an issue is partial updates with PATCH in FastAPI, though I want to point out that my intention was not to discuss specific use cases but to focus on the general (architectural?) problems.

To conclude, I believe:

1. The optionality definition for the fields should be decoupled from the annotation/default value of the field.
2. Such optional fields, when not specified during model initialization, should not be automatically included in the model instance.

When I propose these things, I do realize that I have no idea about the internals of Pydantic, and I understand that there are reasons why it works how it does. Consider this as a request for comments and an indication of interest to work on the matter if it makes sense and is realistic to improve/fix.

**Comment by samuelcolvin** [comment]
The short answer is that we're not going to change this now for V2 - for good or bad, we're matching how dataclasses work, doing anything else today would be massively confusing for most people.

Doing anything else would also make pydantic contradictory to all IDEs and static typing libraries.

However, there's a few factors that should help to mitigate your concerns:

1. `__fields_set__` records which fields are set (e.g. are not default), and optionally only serialize fields which have been set
2. V2 will also have fully support for TypedDict, which supports the idea of missing, as distinct from having a default value.
3. [pydantic-core](https://github.com/pydantic/pydantic-core), which will drive all the validation for V2 is very flexible, so you could always use that build your own model, your own way.
